src/ray/util/shared_lru.h

```diff
@@ -0,0 +1,203 @@
+
+    return value;
+  }
+
+  // Clear the cache.
+  void Clear() {
+    cache_.clear();
+    lru_list_.clear();
+  }
+
+  // Accessors for cache parameters.
+  size_t max_entries() const { return max_entries_; }
+
+ private:
+  struct Entry {
+    // The entry's value.
+    Val value;
+
+    // A list iterator pointing to the entry's position in the LRU list.
+    typename std::list<Key>::iterator lru_iterator;
+  };
+
+  using EntryMap = absl::flat_hash_map<Key, Entry>;
+
+  // The maximum number of entries in the cache. A value of 0 means there is no
+  // limit on entry count.
+  const size_t max_entries_;
+
+  // All keys are stored as refernce (`std::reference_wrapper`), and the
+  // ownership lies in `lru_list_`.
+  EntryMap cache_;
+
+  // The LRU list of entries. The front of the list identifies the most
+  // recently accessed entry.
+  std::list<Key> lru_list_;
+};
+
+// Same interfaces as `SharedLruCache`, but all cached values are
+// `const`-specified to avoid concurrent updates.
+template <typename K, typename V>
+using SharedLruConstCache = SharedLruCache<K, const V>;
+
+// Same interface and functionality as `SharedLruCache`, but thread-safe version.
+template <typename Key, typename Val>
+class ThreadSafeSharedLruCache final {
+ public:
+  using key_type = Key;
+  using mapped_type = Val;
+
+  // A `max_entries` of 0 means that there is no limit on the number of entries
+  // in the cache.
+  explicit ThreadSafeSharedLruCache(size_t max_entries) : cache_(max_entries) {}
+
+  ThreadSafeSharedLruCache(const ThreadSafeSharedLruCache &) = delete;
+  ThreadSafeSharedLruCache &operator=(const ThreadSafeSharedLruCache &) = delete;
+
+  ~ThreadSafeSharedLruCache() = default;
+
+  // Insert `value` with key `key`. This will replace any previous entry with
+  // the same key.
+  void Put(Key key, Val value) {
+    std::lock_guard lck(mtx_);
+    cache_.Put(std::move(key), std::move(value));
+  }
+
+  // Delete the entry with key `key`. Return true if the entry was found for
+  // `key`, false if the entry was not found. In both cases, there is no entry
+  // with key `key` existed after the call.
+  bool Delete(const Key &key) {
+    std::lock_guard lck(mtx_);
+    return cache_.Delete(key);
+  }
+
+  // Look up the entry with key `key`. Return std::nullopt if key doesn't exist.
+  // If found, return a copy for the value.
+  std::optional<Val> Get(const Key &key) {
+    std::lock_guard lck(mtx_);
+    return cache_.Get(key);
+  }
+
+  // Clear the cache.
+  void Clear() {
+    std::lock_guard lck(mtx_);
+    cache_.Clear();
+  }
+
+  // Accessors for cache parameters.
+  size_t max_entries() const { return cache_.max_entries(); }
+
+ private:
+  std::mutex mtx_;
+  SharedLruCache<Key, Val> cache_;
+};
+
+// Same interfaces as `SharedLruCache`, but all cached values are
+// `const`-specified to avoid concurrent updates.
+template <typename K, typename V>
+using ThreadSafeSharedLruConstCache = ThreadSafeSharedLruCache<K, const V>;
```

Comment:
not sure if we need this we don't give direct access to values anyways and const-ing value means we're copying all elems when underlying map has to resize

if we want api that doesn't need to copy value, could have a visitation api for threadsafe like a lot of other concurrent containers, where you can have
```
visit(key, function<void (V&)>) {
    lock(mutex);
    function(elem);
}
```

Reply:
My point is:
- As a cache, the key value pair stored are supposed to be constant; otherwise you suffer undefined behavior and data race;
- The reason I use `Lru<Key, Val>`, which allows non-const value type, is quite most of the call sites doesn't enforce const-ness.

For example, the following code doesn't compile:
```
class MyClass {
 public:
  MyClass() { std::cout << "ctor" << std::endl; }
  MyClass(const MyClass&) { std::cout << "copy" << std::endl; }
  MyClass(MyClass&&) { std::cout << "move" << std::endl; }
};
Cache<Key, std::shared_ptr<const MyClass>> cache;

std::shared_ptr<MyClass> Get() {
  return cache.Get(key);
}
```

There're two reasons const alias is necessary and reasonable:
- It stresses the LRU cache does emphasize const-ness for the value;
- It's useful in cases without propagation, see example below

Intended example:
```
ConstCache<Key, Value> cache; // the cached value is of type `const Value`

void Handle() {
  const auto& value = cache.Get(key);
  // Do something to value without propagation
}
```

Reply:
> if we want api that doesn't need to copy value

I have a question for this interface,
- What if the functor is heavy-lifted? Should we place it inside of critical section? It could lead to deadlock as well
- If we don't put the executable in the lock, we could have iterator invalid, since functor takes a reference instead of value, IIUC
- There's one possible way, which is use shared pointer to wrap around all values, but that leads to hidden overhead for all operations (extra memory allocation at Put, pointer chasing at Get)

Reply:
> > if we want api that doesn't need to copy value
> 
> I have a question for this interface,
> 
https://www.boost.org/doc/libs/1_86_0/libs/unordered/doc/html/unordered.html#concurrent_visitation_based_api
boost's description of visitation api for concurrent containers, functor executes inside lock, up to user whether to use it or copy out depending on cost of functor

> There're two reasons const alias is necessary and reasonable:
can talk about this on a quick huddle, i'm thinking constness is up to the public api offered and actually typing as const isn't worth the perf hit from having const types in containers that can reallocate under the hood

Reply:
> But, in general, visitation functions should be as lightweight as possible to reduce contention and increase parallelization.

I think the interface has strong assumption.

I provide such feature in the similar interface in my prev company, but find out not all the callers are aware of the limitation, which leads to threads get blocked.
(and I have to implement another monitor thread to detect long lock held threads...)

Reply:
oh wow that's rough, but I don't think we should not include something because people may not use it correctly even if it is properly commented, but ok for not adding rn now if copying out makes sense in runtime env use cases currently

Reply:
I updated the value type to be shared pointer, should be good now.